scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/Cast.scala

```diff
@@ -0,0 +1,62 @@
+package pl.touk.nussknacker.engine.extension
+
+import cats.data.ValidatedNel
+import cats.implicits.catsSyntaxValidatedId
+import pl.touk.nussknacker.engine.api.Documentation
+import pl.touk.nussknacker.engine.api.generics.{GenericFunctionTypingError, GenericType, TypingFunction}
+import pl.touk.nussknacker.engine.api.typed.typing
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedObjectWithValue}
+
+import scala.util.{Failure, Success, Try}
+
+sealed trait Cast {
+
+  @Documentation(description = "Checks if a type can be casted to a given class")
+  def canCastTo(clazzType: String): Boolean
+
+  @Documentation(description = "Casts a type to a given class or throws exception if type cannot be casted.")
+  @GenericType(typingFunction = classOf[CastTyping.Typing])
+  def castTo[T](clazzType: String): T
+
+}
+
+object CastTyping {
+
+  class Typing extends TypingFunction {
+
+    override def computeResultType(
+        arguments: List[typing.TypingResult]
+    ): ValidatedNel[GenericFunctionTypingError, typing.TypingResult] = arguments match {
+      case TypedObjectWithValue(_, clazzName: String) :: Nil =>
+        Try(Class.forName(clazzName)) match {
+          case Success(clazz) => Typed.typedClass(clazz).validNel
```

Comment:
We should check if casting is possible during typing - e.g. we shouldn't allow to `"some string".castTo("Long")`. I see that we have something like `TypingFunctionForClassMember` for that purpose

Reply:
Theres also some security concerns if we dont do some check.

1. You can infer what types are on classPath - for example: `{}.castTo("pl.touk.nussknacker.ui.server.NuAkkaHttpServerInterpreterForTapirPurposes")` will return type `NuAkkaHttpServerInterpreterForTapirPurposes
2. You can cast to types which allow for operations we shouldn't allow. For example `#input.myMap.castTo("java.util.HashMap").remove("key1")` will do a mutation.

Reply:
I think that `.remove` won't work, because this method wasn't discovered for `HashMap` type, but `Class.forName` with any class available in the classpath indeed looks like something potentially malicious (but currently I don't know how it can be used). I have one more notice - I think that the classloader can be picked incorrectly, see the implementation in JRE:
```
        Class<?> caller = Reflection.getCallerClass();
        return forName0(className, true, ClassLoader.getClassLoader(caller), caller);
```

Reply:
> I think that the classloader can be picked incorrectly, see the implementation in JRE|
It's true, I noticed it yesterday.
> I think that .remove won't work, because this method wasn't discovered for HashMap type
This is true
> with any class available in the classpath indeed looks like something potentially malicious 
This is also true, we can restrict the classes to what `ClassDiscovery` returns but I wonder whether this is enough.

Reply:
Run this test in SpelExpressionSpec:
```scala
test("debug") {
    import scala.collection.JavaConverters._
    val newctx = ctx.withVariable("hashMap", new java.util.HashMap[String, Int](Map("meh" -> 2).asJava))
    val result =
      parse[Any]("#hashMap.castTo('java.util.HashMap').remove('meh')", newctx).validExpression.evaluateSync[Any](newctx)
    println(result)
  }
```
The problem seems to be here https://github.com/TouK/nussknacker/blob/3e70f0020b115816b8aeb3801426957ba615f5f5/scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/spel/typer/MethodReferenceTyper.scala#L38 - returns None and then its mapped to Right(Unknown) and gets executed. I think this is the case for all classes which have no classDefinition

Reply:
After some change locally I have:
```
Invalid(NonEmptyList(GenericFunctionError(java.util.HashMap is not allowed), IllegalInvocationError(Unknown)))
```